### ISSUE TITLE
Probability for applying an augmentation

### ISSUE THREAD

**Issue by shijianjian** [kornia/kornia#656]
## 🚀 Feature
@edgarriba @pmeier  
As discussed in Slack, we are thinking of enabling the ```p``` for all augmentation methods. I have drafted a sample code for discussion here. The code maintains the reproducibility and would simplify the ```apply_transform``` logic. Specifically, I have added ```forward_parameters``` to add the boolean mask and ```forward_transform``` to apply the transformation.

The most significant breaking change here is the ```generate_parameters``` must return a dict like:
```python
{
  "params": {"crop_src": ..., "crop_dst": ...},
  "flags": {"align_corners": True, "interpolation": 0}
}
```
Also, it gonna be hard to maintain the functional functions like ```kornia.augmentation.random_crop()```. I am thinking of either to deprecate those functions or to make functions like ```forward_transform``` and ```forward_parameters``` static. Then we may implement those easier without initiating a ```nn.Module```. e.g. something like:
```python
def random_crop(input, ...):
    params = RandomCrop.forward_parameters(...)
    return RandomCrop.forward_transform(input, params)
```

Below is the implementation:
```python
class AugmentationBase(nn.Module):
    r"""AugmentationBase base class for customized augmentation implementations. For any augmentation,
    the implementation of "generate_parameters" and "apply_transform" are required while the
    "compute_transformation" is only required when passing "return_transform" as True.

    Args:
        p (float): probability for applying an augmentation. This param controls the augmentation probabilities
                   element-wisely for a batch.
        return_transform (bool): if ``True`` return the matrix describing the geometric transformation applied to each
                                      input tensor. If ``False`` and the input is a tuple the applied transformation
                                      wont be concatenated.
        same_on_batch (bool): apply the same transformation across the batch. Default: False
    """

    def __init__(self, p: float, return_transform: bool = False, same_on_batch: bool = False) -> None:
        super(AugmentationBase, self).__init__()
        self.p = p
        self.return_transform = return_transform
        self.same_on_batch = same_on_batch

    def __repr__(self) -> str:
        return f"p={self.p}, return_transform={self.return_transform}, same_on_batch={self.same_on_batch}"

    def infer_batch_shape(self, input: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]) -> torch.Size:
        return _infer_batch_shape(input)

    def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor]:
        raise NotImplementedError

    def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
        raise NotImplementedError

    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
        raise NotImplementedError

    def forward_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor]:
        params = self.generate_parameters(batch_shape)
        batch_probs = rg.random_prob_generator(batch_shape[0], self.p, self.same_on_batch)
        params.update(batch_probs)
        return params

    def _forward_input(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
        """ Forward the transformation.

        Args:
            params (Dict[str, torch.Tensor]): A dict with the following keys.
                - params['batch_prob']: element-wise boolean mask to determin if to apply the augmentation or not.
                - params['params']: element-wise param for each element in a batch.
                - params['flags']: overall config flags for the augmentation
        """
        input = _transform_input(input)
        to_apply = params['batch_prob']
        _validate_input_dtype(input, accepted_dtypes=[torch.float16, torch.float32, torch.float64])

        # select the inputs to be augmented
        transformed = input.clone()
        # select the params to be augmented
        for k, v in trainable_params['params'].items():
            trainable_params.update(k, v[to_apply])
        params = trainable_params['params'].copy()
        params.update(static_params['flags'])
        transformed[to_apply] = self.apply_transform(input[to_apply], params)
        return transformed

    def forward_transform(self, input: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]],
                          params: Optional[Dict[str, torch.Tensor]] = None,  # type: ignore
                          return_transform: Optional[bool] = None
                          ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:  # type: ignore
        if isinstance(input, tuple):
            output = self._forward_input(input[0], self._params)
            transformation_matrix = self.compute_transformation(input[0], self._params)
            if return_transform:
                return output, transformation_matrix @ input[1]
            else:
                return output, input[1]

        output = self._forward_input(input, self._params)
        if return_transform:
            transformation_matrix = self.compute_transformation(input, self._params)
            return output, transformation_matrix
        return output

    def forward(self, input: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]],
                params: Optional[Dict[str, torch.Tensor]] = None,  # type: ignore
                return_transform: Optional[bool] = None
                ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:  # type: ignore
        if return_transform is None:
            return_transform = self.return_transform
        if params is None:
            batch_shape = self.infer_batch_shape(input)
            params = self.forward_parameters(batch_shape)
        self._params = params

        return self.forward_transform(input, self._params, return_transform)
```

## Motivation

To control if to apply a transformation or not.

**Comment by shijianjian** [comment]
closed by #664
